engines/freescape/freescape.cpp

```diff
@@ -0,0 +1,609 @@
+	event.mouse = Common::Point(mouseX, mouseY);
+	event.customType = 0xde00;
+	g_system->getEventManager()->pushEvent(event);
+
+	byte nextKeyCode = _demoData[_demoIndex++];
+
+	if (nextKeyCode == 0x30) {
+		event.type = Common::EVENT_LBUTTONDOWN; // Keep same event fields
+		g_system->getEventManager()->pushEvent(event);
+		nextKeyCode = _demoData[_demoIndex++];
+	} else {
+		while (nextKeyCode != 0) {
+			event = Common::Event();
+			event.type = Common::EVENT_KEYDOWN;
+			event.kbd.keycode = (Common::KeyCode)decodeAmigaAtariKey(nextKeyCode);
+			debugC(1, kFreescapeDebugMove, "Pushing key: %x", event.kbd.keycode);
+			event.customType = 0xde00;
+			g_system->getEventManager()->pushEvent(event);
+			nextKeyCode = _demoData[_demoIndex++];
+		}
+	}
+	assert(!nextKeyCode);
+	g_system->delayMillis(50);
+}
+
+void FreescapeEngine::processInput() {
+	float currentFrame = g_system->getMillis();
+	float deltaTime = 20.0;
+	_lastFrame = currentFrame;
+	Common::Event event;
+	Common::Point mousePos;
+	while (g_system->getEventManager()->pollEvent(event)) {
+		if (_demoMode) {
+			if (event.customType != 0xde00)
+				continue;
+		}
+
+		switch (event.type) {
+		case Common::EVENT_KEYDOWN:
+			if (event.kbd.keycode == Common::KEYCODE_o || event.kbd.keycode == Common::KEYCODE_UP)
+				move(kForwardMovement, _scaleVector.x(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_k || event.kbd.keycode == Common::KEYCODE_DOWN)
+				move(kBackwardMovement, _scaleVector.x(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_LEFT)
+				move(kLeftMovement, _scaleVector.y(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_RIGHT)
+				move(kRightMovement, _scaleVector.y(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_KP5 || event.kbd.keycode == Common::KEYCODE_KP0)
+				shoot();
+			else if (event.kbd.keycode == Common::KEYCODE_p)
+				rotate(0, 5);
+			else if (event.kbd.keycode == Common::KEYCODE_l)
+				rotate(0, -5);
+			else if (event.kbd.keycode == Common::KEYCODE_u)
+				rotate(180, 0);
+			else if (event.kbd.keycode == Common::KEYCODE_q)
+				rotate(-_angleRotations[_angleRotationIndex], 0);
+			else if (event.kbd.keycode == Common::KEYCODE_w)
+				rotate(_angleRotations[_angleRotationIndex], 0);
+			else if (event.kbd.keycode == Common::KEYCODE_r)
+				rise();
+			else if (event.kbd.keycode == Common::KEYCODE_f)
+				lower();
+			else if (event.kbd.keycode == Common::KEYCODE_n) {
+				_noClipMode = !_noClipMode;
+				_flyMode = _noClipMode;
+			} else if (event.kbd.keycode == Common::KEYCODE_ESCAPE)
+				openMainMenuDialog();
+			else if (event.kbd.keycode == Common::KEYCODE_SPACE) {
+				_shootMode = !_shootMode;
+				if (!_shootMode) {
+					_crossairPosition.x = _screenW / 2;
+					_crossairPosition.y = _screenH / 2;
+				}
+			} else
+				pressedKey(event.kbd.keycode);
+			break;
+
+		case Common::EVENT_QUIT:
+		case Common::EVENT_RETURN_TO_LAUNCHER:
+			quitGame();
+			return;
+			break;
```

Comment:
This line is unreachable

Reply:
Isn't the C++ compiler complaining if you don't add break, even with a return?

Reply:
on the contrary, it knows that `return` does not continue and the advanced compiler will complain about the unreachable code.

Reply:
Removed!